feat: importNamePattern option in no-restricted-imports (#17721)
eslint/eslint: docs/src/rules/no-restricted-imports.md

}]
```

Regex patterns can also be used to restrict specific import Name:

```json
"no-restricted-imports": ["error", {
    "patterns": [{
      "group": ["import-foo/*"],
      "importNamePattern": "^foo",
    }]
}]
```

> **Note:** If `importNames` and `importNamePattern` are used together then the `importNames` will be first considered.

Reviewer: ```suggestion
**Note:** If `importNames` and `importNamePattern` are used together, it initially reviews the value specified in `importNames`. Afterward, it focuses on assessing the value specified in `importNamePattern`.
```

> **Note:** In patterns array `importNames` and `importNamePatterns` can not be used together. If they are used together then the `importNames` will be considered.

Reviewer 2: We didn't discuss this on the issue, but I don't think there's a reason not to allow both to be used at the same time.

Author: Agreed, 
Then can i get an overview about how they should work together and what if they are referring the same import name.

Reviewer 3: @mdjermanovic is away for the rest of the week, but here's my thinking: you should first check the name against all `importNames` and if there aren't any matches, then check against `importNamePattern`.

Author: Sounds good to me, working on this approach.

> **Note:** If `importNames` and `importNamePattern` are used together, it initially reviews the value specified in `importNames`. Afterward, it focuses on assessing the value specified in `importNamePattern`.

Author: > I'm not sure if this is important to note in the docs as it seems like a technical detail that doesn't affect the outcome.

as i am getting this line you want me to remove just this `note` tag,
or the whole statement?

> **Note:** If `importNames` and `importNamePattern` are used together, it initially reviews the value specified in `importNames`. Afterward, it focuses on assessing the value specified in `importNamePattern`.

Reviewer 2: I think we can remove the whole statement.

> **Note:** If `importNames` and `importNamePattern` are used together, it initially reviews the value specified in `importNames`. Afterward, it focuses on assessing the value specified in `importNamePattern`.

Author: Done!

import { isEmpty } from 'utils/collection-utils';
```

> ::: incorrect { "sourceType": "module" }

Reviewer 2: Looks like the closing `:::` for the example above is now missing.

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["utils/*"],
    importNamePattern: '^is',

>     message: "Use 'isEmpty' from lodash instead."

Reviewer 2: ```suggestion
    message: "Use 'is*' functions from lodash instead."
```

I think this message would make more sense for this example as the configuration doesn't disallow only "isEmpty" but any names that start with "is".

}]}]*/

import { isEmpty } from 'utils/collection-utils';
```

:::

::: incorrect { "sourceType": "module" }

```js

> /*eslint no-restricted-imports: ["error", { patterns: [{
>     group: ["foo/*"],
>     importNamePattern: '^(is|bar)',
>     message: "Use 'isBar' from baz/bar instead"
> }]}]*/
> 
> import bar, { isBar } from 'foo/bar';

Reviewer 2: This example is confusing because `bar` matches the pattern but is allowed since it's just a local name.

Maybe something like this:

```suggestion
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["foo/*"],
    importNamePattern: '^(is|has)',
    message: "Use 'is*' and 'has*' functions from baz/bar instead"
}]}]*/

import { isSomething, hasSomething } from 'foo/bar';
```

```

:::

::: incorrect { "sourceType": "module" }

```js

> /*eslint no-restricted-imports: ["error", { patterns: [{
>     importNames: ["foo"],
>     group: ["foo/*"],
>     importNamePattern: '^bar',
> }]}]*/
> 
> import { bar } from 'foo/bar';

Reviewer 2: ```suggestion
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["foo/*"],
    importNames: ["bar"],
    importNamePattern: '^baz',
}]}]*/

import { bar, bazQux } from 'foo/quux';
```

To clarify that both apply.

import { isEmpty } from 'utils/collection-utils';
```

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["utils/*"],
    importNamePattern: '^is',
    message: "Use 'isEmpty' from lodash instead."
}]}]*/

import { isEmpty } from 'utils/collection-utils';
```

:::

::: incorrect { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    importNames: ["foo"],
    group: ["bar/*"],
    importNamePattern: '^bar',
}]}]*/

> import { bar } from 'bar/*';

Reviewer: Can we also add an example where we specify multiple patterns (`^(is|bar)`)?

Reviewer: and a test case too

Reviewer: ```suggestion
import { bar } from 'bar/foo';
```

:::

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["utils/*"],
    importNamePattern: '^is',
    message: "Use 'isEmpty' from lodash instead."
}]}]*/

> import isEmpty, { hasValue } from 'utils/collection-utils';

Reviewer: `isEmpty` would not be reported?

:::

::: correct { "sourceType": "module" }

```js
/*eslint no-restricted-imports: ["error", { patterns: [{
    group: ["utils/*"],
    importNamePattern: '^is',

>     message: "Use 'isEmpty' from lodash instead."

Reviewer 2: ```suggestion
    message: "Use 'is*' functions from lodash instead."
```
